File: backend/src/main/java/fr/sylvainjanet/tracker/journal/domain/WeightMeasurement.java
package fr.sylvainjanet.tracker.journal.domain;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Objects;

public final class WeightMeasurement {

    private final LocalDate date;
    private final Weight weight;

    private WeightMeasurement(LocalDate date, Weight weight) {
        this.date = Objects.requireNonNull(date, "date must not be null");
        this.weight = Objects.requireNonNull(weight, "weight must not be null");
    }

    public static WeightMeasurement create(LocalDate date, Weight weight) {
        return new WeightMeasurement(date, weight);
    }

    public LocalDate date() {
        return date;
    }

    public Weight weight() {
        return weight;
    }

    public BigDecimal weightInKilograms() {
        return weight.inKilograms();
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        WeightMeasurement that = (WeightMeasurement) o;
        return Objects.equals(date, that.date);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(date);
    }
}

File: frontend/src/app/shared/domain/weight-measurement.ts
import { type Weight } from './weight';
import type { CalendarDate } from './calendar-date';

export class WeightMeasurement {
  constructor(
    readonly calendarDate: CalendarDate,
    readonly weight: Weight,
  ) {
    if (!calendarDate) {
      throw new Error('calendarDate is required');
    }
    if (!weight) {
      throw new Error('weight is required');
    }
  }

  public static create(calendarDate: CalendarDate, weight: Weight): WeightMeasurement {
    return new WeightMeasurement(calendarDate, weight);
  }

  public weightInKilograms(): number {
    return this.weight.inKilograms();
  }
}
